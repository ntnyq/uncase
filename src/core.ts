import {
  camelCase,
  capitalCase,
  constantCase,
  dotCase,
  kebabCase,
  noCase,
  pascalCase,
  pascalSnakeCase,
  pathCase,
  sentenceCase,
  snakeCase,
  trainCase,
} from 'change-case'
import type { Options } from 'change-case'

/**
 * Type for all availabel [change-case](https://github.com/blakeembrey/change-case) method in `camelCase`
 */
export type CaseType =
  | 'camelCase'
  | 'capitalCase'
  | 'constantCase'
  | 'dotCase'
  | 'kebabCase'
  | 'noCase'
  | 'pascalCase'
  | 'pascalSnakeCase'
  | 'pathCase'
  | 'sentenceCase'
  | 'snakeCase'
  | 'trainCase'

/**
 * Case convert result
 */
export type CaseConvertResult = {
  /**
   * input value
   */
  input: string

  /**
   * whether output has changed from input
   */
  changed: boolean

  /**
   * converted value
   */
  output: string
}

/**
 * Case converter
 */
export type CaseConverter = (input: string) => CaseConvertResult

/**
 * CaseType to case convert method map
 */
export const convertersMap: Record<
  CaseType,
  (value: string, options?: Options) => string
> = {
  camelCase,
  capitalCase,
  constantCase,
  dotCase,
  kebabCase,
  noCase,
  pascalCase,
  pascalSnakeCase,
  pathCase,
  sentenceCase,
  snakeCase,
  trainCase,
}

/**
 * Get a converter by caseType and convert the given input
 *
 * @param caseType - case utils name
 * @param options - case utils options
 * @returns case convert result {@link CaseConvertResult}
 *
 * @see {@link https://github.com/blakeembrey/change-case}
 *
 * @example
 *
 * ```ts
 * import { getCaseConverter } from 'uncase'
 *
 * const result = getCaseConverter('camelCase')('hello-world')
 *
 * console.log(result)
 * // => { input: 'hello-world', changed: true, output: 'helloWorld' }
 * ```
 */
export function getCaseConverter(
  caseType: CaseType,
  options: Options = {},
): CaseConverter {
  const convert = convertersMap[caseType]

  return (input: string) => {
    const output = convert(input, options)
    const changed = output !== input
    return {
      input,
      changed,
      output,
    }
  }
}
