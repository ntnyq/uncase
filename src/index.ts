import {
  camelCase,
  capitalCase,
  constantCase,
  dotCase,
  kebabCase,
  noCase,
  pascalCase,
  pascalSnakeCase,
  pathCase,
  sentenceCase,
  snakeCase,
  trainCase,
} from 'change-case'
import type { Options } from 'change-case'

/**
 * availabel change-case utils
 */
export type CaseType =
  | 'camelCase'
  | 'capitalCase'
  | 'constantCase'
  | 'dotCase'
  | 'kebabCase'
  | 'noCase'
  | 'pascalCase'
  | 'pascalSnakeCase'
  | 'pathCase'
  | 'sentenceCase'
  | 'snakeCase'
  | 'trainCase'

/**
 * change-case utils map
 */
export const convertersMap: Record<
  CaseType,
  (value: string, options?: Options) => string
> = {
  camelCase,
  capitalCase,
  constantCase,
  dotCase,
  kebabCase,
  noCase,
  pascalCase,
  pascalSnakeCase,
  pathCase,
  sentenceCase,
  snakeCase,
  trainCase,
}

/**
 * get a converter and convert given input
 *
 * @param caseType - case utils name
 * @param options - case utils options
 * @returns input, output and whether changed
 *
 * @see {@link https://github.com/blakeembrey/change-case}
 */
export function getCaseConverter(caseType: CaseType, options: Options = {}) {
  const convert = convertersMap[caseType]

  return (input: string) => {
    const output = convert(input, options)
    const changed = output !== input
    return {
      input,
      changed,
      output,
    }
  }
}
