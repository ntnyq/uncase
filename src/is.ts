import { getCaseConverter } from '.'
import type { CaseType } from '.'

/**
 * Non-empty string
 */
type NonEmptyString = string & { 0: '' }

/**
 * Case validator
 */
type CaseValidator = (input: string) => boolean

/**
 * Check if given value is a string
 *
 * @param value - given value
 * @returns true if it is a string, false otherwise
 */
function isString(value: unknown): value is string {
  return typeof value === 'string'
}

/**
 * Check if given value is a non-empty string
 *
 * @param value - given value
 * @returns true if it is a non-empty string, false otherwise
 */
function isNonEmptyString(value: unknown): value is NonEmptyString {
  return isString(value) && value.length > 0
}

/**
 * Create a case validator by caseType
 *
 * @param caseType - {@link CaseType}
 * @returns case validator
 */
function createCaseValidator(caseType: CaseType): CaseValidator {
  const convert = getCaseConverter(caseType)
  return (input: string) => {
    if (!isNonEmptyString(input)) return false
    return !convert(input).changed
  }
}

/**
 * Check if given input string is in camelCase
 *
 * @param input - input string
 * @returns true if camelCase, false otherwise
 */
export const isCamelCase = (input: string) =>
  createCaseValidator('camelCase')(input)

/**
 * Check if given input string is in capitalCase
 *
 * @param input - input string
 * @returns true if capitalCase, false otherwise
 */
export const isCapitalCase = (input: string) =>
  createCaseValidator('capitalCase')(input)

/**
 * Check if given input string is in constantCase
 *
 * @param input - input string
 * @returns true if constantCase, false otherwise
 */
export const isConstantCase = (input: string) =>
  createCaseValidator('constantCase')(input)

/**
 * Check if given input string is in dotCase
 *
 * @param input - input string
 * @returns true if dotCase, false otherwise
 */
export const isDotCase = (input: string) =>
  createCaseValidator('dotCase')(input)

/**
 * Check if given input string is in kebabCase
 *
 * @param input - input string
 * @returns true if kebabCase, false otherwise
 */
export const isKebabCase = (input: string) =>
  createCaseValidator('kebabCase')(input)

/**
 * Check if given input string is in noCase
 *
 * @param input - input string
 * @returns true if noCase, false otherwise
 */
export const isNoCase = (input: string) => createCaseValidator('noCase')(input)

/**
 * Check if given input string is in pascalCase
 *
 * @param input - input string
 * @returns true if pascalCase, false otherwise
 */
export const isPascalCase = (input: string) =>
  createCaseValidator('pascalCase')(input)

/**
 * Check if given input string is in pascalSnakeCase
 *
 * @param input - input string
 * @returns true if pascalSnakeCase, false otherwise
 */
export const isPascalSnakeCase = (input: string) =>
  createCaseValidator('pascalSnakeCase')(input)

/**
 * Check if given input string is in pathCase
 *
 * @param input - input string
 * @returns true if pathCase, false otherwise
 */
export const isPathCase = (input: string) =>
  createCaseValidator('pathCase')(input)

/**
 * Check if given input string is in sentenceCase
 *
 * @param input - input string
 * @returns true if sentenceCase, false otherwise
 */
export const isSentenceCase = (input: string) =>
  createCaseValidator('sentenceCase')(input)

/**
 * Check if given input string is in snakeCase
 *
 * @param input - input string
 * @returns true if snakeCase, false otherwise
 */
export const isSnakeCase = (input: string) =>
  createCaseValidator('snakeCase')(input)

/**
 * Check if given input string is in trainCase
 *
 * @param input - input string
 * @returns true if trainCase, false otherwise
 */
export const isTrainCase = (input: string) =>
  createCaseValidator('trainCase')(input)
