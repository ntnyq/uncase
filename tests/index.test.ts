import { expect, it } from 'vitest'
import { getCaseConverter } from '../src'
import type { CaseType } from '../src'

export const CASE_MAP: Record<CaseType, string> = {
  camelCase: 'camelCase',
  kebabCase: 'kebab-case',
  pascalCase: 'PascalCase',
  snakeCase: 'snake_case',
  trainCase: 'Train-Case',
  capitalCase: 'Capital Case',
  constantCase: 'CONSTANT_CASE',
  dotCase: 'dot.case',
  noCase: 'no case',
  pascalSnakeCase: 'Pascal_Snake_Case',
  pathCase: 'path/case',
  sentenceCase: 'Sentence case',
} as const

const keys = Object.keys(CASE_MAP)
const values = Object.values(CASE_MAP)

it.each(keys)(`to %s`, caseType => {
  const convert = getCaseConverter(caseType as CaseType)

  expect(values.map(v => convert(v))).toMatchSnapshot()
})
