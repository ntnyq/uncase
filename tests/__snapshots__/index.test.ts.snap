// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`to camelCase 1`] = `
[
  {
    "changed": false,
    "input": "camelCase",
    "output": "camelCase",
  },
  {
    "changed": true,
    "input": "kebab-case",
    "output": "kebabCase",
  },
  {
    "changed": true,
    "input": "PascalCase",
    "output": "pascalCase",
  },
  {
    "changed": true,
    "input": "snake_case",
    "output": "snakeCase",
  },
  {
    "changed": true,
    "input": "Train-Case",
    "output": "trainCase",
  },
  {
    "changed": true,
    "input": "Capital Case",
    "output": "capitalCase",
  },
  {
    "changed": true,
    "input": "CONSTANT_CASE",
    "output": "constantCase",
  },
  {
    "changed": true,
    "input": "dot.case",
    "output": "dotCase",
  },
  {
    "changed": true,
    "input": "no case",
    "output": "noCase",
  },
  {
    "changed": true,
    "input": "Pascal_Snake_Case",
    "output": "pascalSnakeCase",
  },
  {
    "changed": true,
    "input": "path/case",
    "output": "pathCase",
  },
  {
    "changed": true,
    "input": "Sentence case",
    "output": "sentenceCase",
  },
]
`;

exports[`to capitalCase 1`] = `
[
  {
    "changed": true,
    "input": "camelCase",
    "output": "Camel Case",
  },
  {
    "changed": true,
    "input": "kebab-case",
    "output": "Kebab Case",
  },
  {
    "changed": true,
    "input": "PascalCase",
    "output": "Pascal Case",
  },
  {
    "changed": true,
    "input": "snake_case",
    "output": "Snake Case",
  },
  {
    "changed": true,
    "input": "Train-Case",
    "output": "Train Case",
  },
  {
    "changed": false,
    "input": "Capital Case",
    "output": "Capital Case",
  },
  {
    "changed": true,
    "input": "CONSTANT_CASE",
    "output": "Constant Case",
  },
  {
    "changed": true,
    "input": "dot.case",
    "output": "Dot Case",
  },
  {
    "changed": true,
    "input": "no case",
    "output": "No Case",
  },
  {
    "changed": true,
    "input": "Pascal_Snake_Case",
    "output": "Pascal Snake Case",
  },
  {
    "changed": true,
    "input": "path/case",
    "output": "Path Case",
  },
  {
    "changed": true,
    "input": "Sentence case",
    "output": "Sentence Case",
  },
]
`;

exports[`to constantCase 1`] = `
[
  {
    "changed": true,
    "input": "camelCase",
    "output": "CAMEL_CASE",
  },
  {
    "changed": true,
    "input": "kebab-case",
    "output": "KEBAB_CASE",
  },
  {
    "changed": true,
    "input": "PascalCase",
    "output": "PASCAL_CASE",
  },
  {
    "changed": true,
    "input": "snake_case",
    "output": "SNAKE_CASE",
  },
  {
    "changed": true,
    "input": "Train-Case",
    "output": "TRAIN_CASE",
  },
  {
    "changed": true,
    "input": "Capital Case",
    "output": "CAPITAL_CASE",
  },
  {
    "changed": false,
    "input": "CONSTANT_CASE",
    "output": "CONSTANT_CASE",
  },
  {
    "changed": true,
    "input": "dot.case",
    "output": "DOT_CASE",
  },
  {
    "changed": true,
    "input": "no case",
    "output": "NO_CASE",
  },
  {
    "changed": true,
    "input": "Pascal_Snake_Case",
    "output": "PASCAL_SNAKE_CASE",
  },
  {
    "changed": true,
    "input": "path/case",
    "output": "PATH_CASE",
  },
  {
    "changed": true,
    "input": "Sentence case",
    "output": "SENTENCE_CASE",
  },
]
`;

exports[`to dotCase 1`] = `
[
  {
    "changed": true,
    "input": "camelCase",
    "output": "camel.case",
  },
  {
    "changed": true,
    "input": "kebab-case",
    "output": "kebab.case",
  },
  {
    "changed": true,
    "input": "PascalCase",
    "output": "pascal.case",
  },
  {
    "changed": true,
    "input": "snake_case",
    "output": "snake.case",
  },
  {
    "changed": true,
    "input": "Train-Case",
    "output": "train.case",
  },
  {
    "changed": true,
    "input": "Capital Case",
    "output": "capital.case",
  },
  {
    "changed": true,
    "input": "CONSTANT_CASE",
    "output": "constant.case",
  },
  {
    "changed": false,
    "input": "dot.case",
    "output": "dot.case",
  },
  {
    "changed": true,
    "input": "no case",
    "output": "no.case",
  },
  {
    "changed": true,
    "input": "Pascal_Snake_Case",
    "output": "pascal.snake.case",
  },
  {
    "changed": true,
    "input": "path/case",
    "output": "path.case",
  },
  {
    "changed": true,
    "input": "Sentence case",
    "output": "sentence.case",
  },
]
`;

exports[`to kebabCase 1`] = `
[
  {
    "changed": true,
    "input": "camelCase",
    "output": "camel-case",
  },
  {
    "changed": false,
    "input": "kebab-case",
    "output": "kebab-case",
  },
  {
    "changed": true,
    "input": "PascalCase",
    "output": "pascal-case",
  },
  {
    "changed": true,
    "input": "snake_case",
    "output": "snake-case",
  },
  {
    "changed": true,
    "input": "Train-Case",
    "output": "train-case",
  },
  {
    "changed": true,
    "input": "Capital Case",
    "output": "capital-case",
  },
  {
    "changed": true,
    "input": "CONSTANT_CASE",
    "output": "constant-case",
  },
  {
    "changed": true,
    "input": "dot.case",
    "output": "dot-case",
  },
  {
    "changed": true,
    "input": "no case",
    "output": "no-case",
  },
  {
    "changed": true,
    "input": "Pascal_Snake_Case",
    "output": "pascal-snake-case",
  },
  {
    "changed": true,
    "input": "path/case",
    "output": "path-case",
  },
  {
    "changed": true,
    "input": "Sentence case",
    "output": "sentence-case",
  },
]
`;

exports[`to noCase 1`] = `
[
  {
    "changed": true,
    "input": "camelCase",
    "output": "camel case",
  },
  {
    "changed": true,
    "input": "kebab-case",
    "output": "kebab case",
  },
  {
    "changed": true,
    "input": "PascalCase",
    "output": "pascal case",
  },
  {
    "changed": true,
    "input": "snake_case",
    "output": "snake case",
  },
  {
    "changed": true,
    "input": "Train-Case",
    "output": "train case",
  },
  {
    "changed": true,
    "input": "Capital Case",
    "output": "capital case",
  },
  {
    "changed": true,
    "input": "CONSTANT_CASE",
    "output": "constant case",
  },
  {
    "changed": true,
    "input": "dot.case",
    "output": "dot case",
  },
  {
    "changed": false,
    "input": "no case",
    "output": "no case",
  },
  {
    "changed": true,
    "input": "Pascal_Snake_Case",
    "output": "pascal snake case",
  },
  {
    "changed": true,
    "input": "path/case",
    "output": "path case",
  },
  {
    "changed": true,
    "input": "Sentence case",
    "output": "sentence case",
  },
]
`;

exports[`to pascalCase 1`] = `
[
  {
    "changed": true,
    "input": "camelCase",
    "output": "CamelCase",
  },
  {
    "changed": true,
    "input": "kebab-case",
    "output": "KebabCase",
  },
  {
    "changed": false,
    "input": "PascalCase",
    "output": "PascalCase",
  },
  {
    "changed": true,
    "input": "snake_case",
    "output": "SnakeCase",
  },
  {
    "changed": true,
    "input": "Train-Case",
    "output": "TrainCase",
  },
  {
    "changed": true,
    "input": "Capital Case",
    "output": "CapitalCase",
  },
  {
    "changed": true,
    "input": "CONSTANT_CASE",
    "output": "ConstantCase",
  },
  {
    "changed": true,
    "input": "dot.case",
    "output": "DotCase",
  },
  {
    "changed": true,
    "input": "no case",
    "output": "NoCase",
  },
  {
    "changed": true,
    "input": "Pascal_Snake_Case",
    "output": "PascalSnakeCase",
  },
  {
    "changed": true,
    "input": "path/case",
    "output": "PathCase",
  },
  {
    "changed": true,
    "input": "Sentence case",
    "output": "SentenceCase",
  },
]
`;

exports[`to pascalSnakeCase 1`] = `
[
  {
    "changed": true,
    "input": "camelCase",
    "output": "Camel_Case",
  },
  {
    "changed": true,
    "input": "kebab-case",
    "output": "Kebab_Case",
  },
  {
    "changed": true,
    "input": "PascalCase",
    "output": "Pascal_Case",
  },
  {
    "changed": true,
    "input": "snake_case",
    "output": "Snake_Case",
  },
  {
    "changed": true,
    "input": "Train-Case",
    "output": "Train_Case",
  },
  {
    "changed": true,
    "input": "Capital Case",
    "output": "Capital_Case",
  },
  {
    "changed": true,
    "input": "CONSTANT_CASE",
    "output": "Constant_Case",
  },
  {
    "changed": true,
    "input": "dot.case",
    "output": "Dot_Case",
  },
  {
    "changed": true,
    "input": "no case",
    "output": "No_Case",
  },
  {
    "changed": false,
    "input": "Pascal_Snake_Case",
    "output": "Pascal_Snake_Case",
  },
  {
    "changed": true,
    "input": "path/case",
    "output": "Path_Case",
  },
  {
    "changed": true,
    "input": "Sentence case",
    "output": "Sentence_Case",
  },
]
`;

exports[`to pathCase 1`] = `
[
  {
    "changed": true,
    "input": "camelCase",
    "output": "camel/case",
  },
  {
    "changed": true,
    "input": "kebab-case",
    "output": "kebab/case",
  },
  {
    "changed": true,
    "input": "PascalCase",
    "output": "pascal/case",
  },
  {
    "changed": true,
    "input": "snake_case",
    "output": "snake/case",
  },
  {
    "changed": true,
    "input": "Train-Case",
    "output": "train/case",
  },
  {
    "changed": true,
    "input": "Capital Case",
    "output": "capital/case",
  },
  {
    "changed": true,
    "input": "CONSTANT_CASE",
    "output": "constant/case",
  },
  {
    "changed": true,
    "input": "dot.case",
    "output": "dot/case",
  },
  {
    "changed": true,
    "input": "no case",
    "output": "no/case",
  },
  {
    "changed": true,
    "input": "Pascal_Snake_Case",
    "output": "pascal/snake/case",
  },
  {
    "changed": false,
    "input": "path/case",
    "output": "path/case",
  },
  {
    "changed": true,
    "input": "Sentence case",
    "output": "sentence/case",
  },
]
`;

exports[`to sentenceCase 1`] = `
[
  {
    "changed": true,
    "input": "camelCase",
    "output": "Camel case",
  },
  {
    "changed": true,
    "input": "kebab-case",
    "output": "Kebab case",
  },
  {
    "changed": true,
    "input": "PascalCase",
    "output": "Pascal case",
  },
  {
    "changed": true,
    "input": "snake_case",
    "output": "Snake case",
  },
  {
    "changed": true,
    "input": "Train-Case",
    "output": "Train case",
  },
  {
    "changed": true,
    "input": "Capital Case",
    "output": "Capital case",
  },
  {
    "changed": true,
    "input": "CONSTANT_CASE",
    "output": "Constant case",
  },
  {
    "changed": true,
    "input": "dot.case",
    "output": "Dot case",
  },
  {
    "changed": true,
    "input": "no case",
    "output": "No case",
  },
  {
    "changed": true,
    "input": "Pascal_Snake_Case",
    "output": "Pascal snake case",
  },
  {
    "changed": true,
    "input": "path/case",
    "output": "Path case",
  },
  {
    "changed": false,
    "input": "Sentence case",
    "output": "Sentence case",
  },
]
`;

exports[`to snakeCase 1`] = `
[
  {
    "changed": true,
    "input": "camelCase",
    "output": "camel_case",
  },
  {
    "changed": true,
    "input": "kebab-case",
    "output": "kebab_case",
  },
  {
    "changed": true,
    "input": "PascalCase",
    "output": "pascal_case",
  },
  {
    "changed": false,
    "input": "snake_case",
    "output": "snake_case",
  },
  {
    "changed": true,
    "input": "Train-Case",
    "output": "train_case",
  },
  {
    "changed": true,
    "input": "Capital Case",
    "output": "capital_case",
  },
  {
    "changed": true,
    "input": "CONSTANT_CASE",
    "output": "constant_case",
  },
  {
    "changed": true,
    "input": "dot.case",
    "output": "dot_case",
  },
  {
    "changed": true,
    "input": "no case",
    "output": "no_case",
  },
  {
    "changed": true,
    "input": "Pascal_Snake_Case",
    "output": "pascal_snake_case",
  },
  {
    "changed": true,
    "input": "path/case",
    "output": "path_case",
  },
  {
    "changed": true,
    "input": "Sentence case",
    "output": "sentence_case",
  },
]
`;

exports[`to trainCase 1`] = `
[
  {
    "changed": true,
    "input": "camelCase",
    "output": "Camel-Case",
  },
  {
    "changed": true,
    "input": "kebab-case",
    "output": "Kebab-Case",
  },
  {
    "changed": true,
    "input": "PascalCase",
    "output": "Pascal-Case",
  },
  {
    "changed": true,
    "input": "snake_case",
    "output": "Snake-Case",
  },
  {
    "changed": false,
    "input": "Train-Case",
    "output": "Train-Case",
  },
  {
    "changed": true,
    "input": "Capital Case",
    "output": "Capital-Case",
  },
  {
    "changed": true,
    "input": "CONSTANT_CASE",
    "output": "Constant-Case",
  },
  {
    "changed": true,
    "input": "dot.case",
    "output": "Dot-Case",
  },
  {
    "changed": true,
    "input": "no case",
    "output": "No-Case",
  },
  {
    "changed": true,
    "input": "Pascal_Snake_Case",
    "output": "Pascal-Snake-Case",
  },
  {
    "changed": true,
    "input": "path/case",
    "output": "Path-Case",
  },
  {
    "changed": true,
    "input": "Sentence case",
    "output": "Sentence-Case",
  },
]
`;
